src/algebra/foo.lean

```diff
@@ -0,0 +1,65 @@
+/-
+Copyright (c) 2021 Eric Rodriguez. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Rodriguez
+-/
+import algebra.algebra.basic
+
+/-!
+# Name Of Class (tba)
+
+We create a typeclass `foo R n` which carries around the fact that `(n : R) ≠ 0`.
+This fact can be really annoying to carry and deduce automatically.
+
+## Main declarations
+
+* `foo R n`: `(n : R) ≠ 0` as a typeclass.
+
+## Implementation details
+
+Does the argument order matter? I would prefer `foo n R`, but not sure what TC prefers.
+
+-/
+
+class foo (R) [has_zero R] [has_one R] [has_add R] (n : ℕ) : Prop := (out : (n : R) ≠ 0)
+
+lemma foo.ne (R) [has_zero R] [has_one R] [has_add R] (n : ℕ) [h : foo R n] : (n : R) ≠ 0 := h.out
+
+namespace foo
+
+universe variables u v w
+variables {R : Type u} {S : Type v} {M : Type w} {n m : ℕ} {a b : ℕ+}
+
+section add_monoid_with_one
+
+variables [add_monoid R] [has_one R]
+
+instance pnat [char_zero R] : foo R a := ⟨by exact_mod_cast a.ne_zero⟩
+
+instance succ [char_zero R] : foo R (n.succ) := ⟨by exact_mod_cast n.succ_ne_zero⟩
+
+end add_monoid_with_one
+
+section semiring
+
+variables [non_assoc_semiring R] [non_assoc_semiring S]
+
+instance one [nontrivial R] : foo R 1 := ⟨ne_of_eq_of_ne nat.cast_one one_ne_zero⟩
+
+-- will this instance cause loops/issues with `foo.pnat` and `foo.succ`?
+instance char_zero [char_zero R] [foo S n] : foo R n :=
```

Comment:
This is a very strange statement. `S` is not needed at all, we just nee `0 < n`.

Reply:
I would just state it for `n : pnat`, otherwise you get a loop going back and forth between `R` and `S`.

Reply:
how else do we provide it without an instance of `foo`? `foo \N n`?

Reply:
I mean something like
```
instance char_zero [char_zero R] : foo R a :=
⟨nat.cast_ne_zero.mpr a.pos.ne'⟩
```

Reply:
It is surely impossible to have both an instance going from `R` to `S` and one going from `S` to `R`.

Reply:
This instance isn't allowed because `S` does not appear on the RHS.